libvips/foreign/jxlsave.c

```diff
@@ -114,65 +122,207 @@ typedef struct _VipsForeignSaveJxl {
+#endif /*DEBUG*/
+
+	VipsForeignSaveJxl *jxl = (VipsForeignSaveJxl *) opaque;
+
+	*format = jxl->format;
+}
+
+static void
+vips_foreign_save_jxl_extra_pixel_format(void *opaque,
+	size_t ec_index, JxlPixelFormat *format)
+{
+#ifdef DEBUG
+	printf("vips_foreign_save_jxl_extra_pixel_format:\n");
+#endif /*DEBUG*/
+
+	return vips_foreign_save_jxl_pixel_format(opaque, format);
+}
+
+static const void *
+vips_foreign_save_jxl_data_at(void *opaque,
+	size_t xpos, size_t ypos, size_t xsize, size_t ysize,
+	size_t *row_offset)
+{
+	VipsForeignSave *save = (VipsForeignSave *) opaque;
+	VipsForeignSaveJxl *jxl = (VipsForeignSaveJxl *) opaque;
+
+#ifdef DEBUG
+	printf("vips_foreign_save_jxl_data_at: "
+		"left = %zd, top = %zd, width = %zd, height = %zd\n",
+		xpos, ypos, xsize, ysize);
+#endif /*DEBUG*/
+
+	VipsImage *tile;
+	if (vips_crop(jxl->page, &tile, xpos, ypos, xsize, ysize, NULL)) {
+		jxl->error = TRUE;
+		return NULL;
+	}
+
+	// disable progress reporting from this copy_memory()
+	vips_image_set_int(tile, "hide-progress", 1);
+
+	VipsImage *memory;
+	if (!(memory = vips_image_copy_memory(tile))) {
+		VIPS_UNREF(tile);
+		jxl->error = TRUE;
+		return NULL;
+	}
+	VIPS_UNREF(tile);
+
+	VipsPel *pels = VIPS_IMAGE_ADDR(memory, 0, 0);
+	*row_offset = VIPS_IMAGE_SIZEOF_LINE(memory);
+
+	g_mutex_lock(&jxl->tile_lock);
+
+	g_assert(!g_hash_table_lookup(jxl->tile_hash, pels));
+	g_hash_table_insert(jxl->tile_hash, pels, memory);
+
+	g_mutex_unlock(&jxl->tile_lock);
+
+	/* Trigger any eval callbacks on our source image and
+	 * check for errors.
+	 *
+	 * FIXME ... can we cancel save with NULL? or will this crash?
+	 */
+	guint64 processed =
+		(guint64) ypos * save->ready->Xsize + (guint64) xpos * ysize;
+	vips_image_eval(save->ready, processed);
+	if (vips_image_iskilled(save->ready))
+		return NULL;
+
+	return pels;
+}
+
+static const void *
+vips_foreign_save_jxl_extra_data_at(void* opaque, size_t ec_index,
+	size_t xpos, size_t ypos, size_t xsize, size_t ysize, size_t* row_offset)
+{
+#ifdef DEBUG
+	printf("vips_foreign_save_jxl_extra_data_at:\n");
+#endif /*DEBUG*/
+
+	return vips_foreign_save_jxl_data_at(opaque,
+		xpos, ypos, xsize, ysize, row_offset);
+}
+
+static void
+vips_foreign_save_jxl_input_release_buffer(void *opaque, const void *pels)
+{
+	VipsForeignSaveJxl *jxl = (VipsForeignSaveJxl *) opaque;
+
+	g_assert(g_hash_table_lookup(jxl->tile_hash, pels));
+	g_hash_table_remove(jxl->tile_hash, pels);
+}
+
+static void
+vips_foreign_save_jxl_set_input_source(VipsForeignSaveJxl *jxl)
+{
+	jxl->input_source = (struct JxlChunkedFrameInputSource) {
+		.opaque = jxl,
+		.get_color_channels_pixel_format = vips_foreign_save_jxl_pixel_format,
+		.get_color_channel_data_at = vips_foreign_save_jxl_data_at,
+		.get_extra_channel_pixel_format =
+			vips_foreign_save_jxl_extra_pixel_format,
+		.get_extra_channel_data_at = vips_foreign_save_jxl_extra_data_at,
+		.release_buffer = vips_foreign_save_jxl_input_release_buffer
+	};
```

Comment:
```suggestion
	};

	g_mutex_init(&jxl->tile_lock);
```

Reply:
I think we can skip the init, since object members are inited to 0. You're right about the _clear though, oops.

Reply:
Ah, you're right. It looks like GLib uses `g_malloc0()` for these structures.
https://github.com/GNOME/glib/blob/2.84.1/gobject/gtype.c#L407

I had assumed otherwise when I made commit 7e2b786c431c9a0a39ea80ffa6afa42e5df5589d, so we can likely drop many of the `g_mutex_init()` calls now. Nice!